### PR TITLE
Full ADSR + Tone.Context

### PR DESCRIPTION
**Pull request by ExSilencio**

### CODE REVIEW

**Review by ExSilencio — COMMENTED** [review]
Here you go.

File: src/app/components/settings-area/settings-area.component.ts [hunk 1]
```diff
@@ -2,6 +2,7 @@ import { Component, OnInit } from '@angular/core';
 import { ADSR,Preset } from 'src/app/Preset';
 import { AudioService } from 'src/app/services/audio.service';
 import { PresetService } from 'src/app/services/preset.service';
+import { KeyComponent } from '../key/key.component';
```
Comment (Maitre-Pangolin) [inline src/app/components/settings-area/settings-area.component.ts]: Remove this

File: src/app/components/settings-area/settings-area.component.ts [hunk 2]
```diff
@@ -52,7 +44,46 @@ waveForms=[{name:'sine',symbol:String.fromCharCode(0x223F)},{name:'square',symbo
...
+        min:0,
+        max:1,
+        step:0.1,
+        value:this.preset.adsr.sustain,
+        action:()=> {
+          this.audioService.setDecay(this.propSliderSustain.value)
+          this.preset.adsr.sustain=this.propSliderSustain.value
+          if(this.propSliderSustain.value == 0){
+            this.keycomp.stopNote();
+          }
```
Comment (Maitre-Pangolin) [inline src/app/components/settings-area/settings-area.component.ts]: Remove this

File: src/app/components/settings-area/settings-area.component.ts [hunk 3]
```diff
@@ -52,7 +44,46 @@ waveForms=[{name:'sine',symbol:String.fromCharCode(0x223F)},{name:'square',symbo
...
+        value:this.preset.adsr.decay,
+        action:()=> {
+          this.audioService.setDecay(this.propSliderDecay.value)
+          this.preset.adsr.decay=this.propSliderDecay.value
+        }
+      }
+
+      this.propSliderSustain={
+        name:'Sustain',
+        min:0,
```
Comment (Maitre-Pangolin) [inline src/app/components/settings-area/settings-area.component.ts:61]: set min to 0.1 or something else just not 0

File: src/app/components/settings-area/settings-area.component.ts [hunk 4]
```diff
@@ -52,7 +44,46 @@ waveForms=[{name:'sine',symbol:String.fromCharCode(0x223F)},{name:'square',symbo
...
+        max:1,
+        step:0.1,
+        value:this.preset.adsr.sustain,
+        action:()=> {
+          this.audioService.setDecay(this.propSliderSustain.value)
```
Comment (Maitre-Pangolin) [inline src/app/components/settings-area/settings-area.component.ts]: Error on method call 

```suggestion
          this.audioService.setSustain(this.propSliderSustain.value)
```

File: src/app/components/keyboard/style.css [hunk 5]
```diff
@@ -0,0 +1,416 @@
...
+    margin-left: 726px;
+    background-color:black;
+    color: white;
+    border-color: black;
+    border-color: red;
+    cursor: pointer;
+    left: 353px;
+    top: 142px;
+}
+
```
Comment (Maitre-Pangolin) [inline src/app/components/keyboard/style.css:406]: Maybe we can try to shorten all this later

**Review by Maitre-Pangolin — CHANGES_REQUESTED** [review]
Weird stuff going on with package.json / package-lock.json and .gitignore

File: src/app/services/audio.service.ts [hunk 6]
```diff
@@ -22,6 +22,11 @@ synthNotes:SynthNote[]=[];
           this.notes.forEach(note=>{
             const key=note+octave
             
+            const context = new Tone.Context({ latencyHint: "playback" });
+            Tone.setContext(context);
+            console.log(Tone.getContext().latencyHint);
+            console.log(Tone.getContext().destination);
+
```
Comment (Maitre-Pangolin) [inline src/app/services/audio.service.ts]: should be outside of the for Each loop

**Review by Maitre-Pangolin — CHANGES_REQUESTED** [review]

**Review by Maitre-Pangolin — APPROVED** [review]

**Pull request merged**